### PR TITLE
Try to load uws by default, fall back to ws

### PR DESCRIPTION
**Pull request by kapouer**
Fixes: #432.

There are some test failures before/after so i'm not sure what to do with them.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ const REPORTER = 'dot';
 gulp.task('default', ['transpile']);
 
 gulp.task('test', ['nsp', 'lint'], function () {
-  if (parseInt(process.versions.node, 10) < 4 && process.env.EIO_WS_ENGINE === 'uws') {
+  if (parseInt(process.versions.node, 10) < 4 && process.env.EIO_WS_ENGINE === 'ws') {
```
Comment (darrachequesne) [inline gulpfile.js]: Shouldn't it be `!==` here?

**Comment by kapouer** [comment]
Well the message in the gulpfile is wrong now, let me fix it.

File: lib/server.js [hunk 2]
```diff
@@ -64,7 +64,14 @@ function Server (opts) {
   // initialize websocket server
   if (~this.transports.indexOf('websocket')) {
     // keep require('ws') as separate expression for packers (browserify, etc)
-    var WebSocketServer = (this.wsEngine ? require(this.wsEngine) : require('ws')).Server;
+    var wsModule;
+    try {
+      wsModule = require(this.wsEngine);
+    } catch (ex) {
+      this.wsEngine = 'ws';
+      wsModule = require(this.wsEngine);
```
Comment (darrachequesne) [inline lib/server.js]: According to the comment above `// keep require('ws') as separate expression for packers (browserify, etc)`, shouldn't it rather be `wsModule = require('ws');` here?

Related: https://github.com/socketio/engine.io/pull/418

File: package.json [hunk 3]
```diff
@@ -47,10 +47,13 @@
     "mocha": "2.3.4",
     "s": "0.1.1",
     "superagent": "0.15.4",
-    "uws": "0.4.0"
+    "uws": "0.12.0"
```
Comment (darrachequesne) [inline package.json]: Shouldn't the "uws" in dev dependencies be removed?

**Comment by darrachequesne** [comment]
Also, the test for `uws` seems to fail:

image

**Comment by kapouer** [comment]
Forwarded test failure to https://github.com/uWebSockets/uWebSockets/issues/365.

**Comment by alexhultman** [comment]
```javascript
var eioc = require('engine.io-client');
var listen = require('./common').listen;
var expect = require('expect.js');

describe('server', function () {
  it('should allow client reconnect after restarting (ws)', function (done) {
    var opts = { transports: ['websocket'] };
    var engine = listen(opts, function (port) {
      // these two lines make the difference between pass or fail
      engine.httpServer.close();
      engine.httpServer.listen(port);

      var socket = new eioc.Socket('ws://localhost:%d'.s(port), { transports: ['websocket'] });

      engine.once('connection', function (conn) {
        setTimeout(function () {
          conn.close();
        }, 10);
      });

      socket.once('close', function (reason) {
        expect(reason).to.be('transport close');
        done();
      });
    });
  });
});
```

This is the only failing test and it passes if you comment out the .close and .listen. This test is very strange and the httpServer being closed and listened on is not part of uws but engine.io itself. I cannot see why it would fail.

**Comment by alexhultman** [comment]
Adding engine.httpServer.removeAllListeners('close'); before engine.httpServer.close also fixes it. Something is messing up in the close handler of the http server of the engine.io

**Comment by alexhultman** [comment]
I've found the issue. Engine.io calls close on the websocket server whenever anything calls close on the httpServer. Engine.io then expects to reuse this closed server, which is not the case in uws. You need to fix this in Engine.io - you are not allowed to close a server and use it later. I would just remove those two lines since they make no sense to have in the first place.

**Comment by darrachequesne** [comment]
@alexhultman I guess it was implemented like node.js native `net.Server`, where you can `close` the server and then `listen` [again](https://nodejs.org/docs/latest/api/net.html#net_server_listen_port_hostname_backlog_callback). What do you think?

**Comment by alexhultman** [comment]
Engineio controls the webspcket module with handleUpgrade and close. The websocket module doesnt care aboit http listen/close in this case - it only cares about the fact that you call websocket module close followed by handleUpgrade. You cannot use a websocket module after closing it. You have to remove or replace this failing test because it fails due to invalid usage of uws.

**Comment by kapouer** [comment]
@alexhultman thank you for your clear answer.

@darrachequesne the problem happened to be, IMO, in engine.io, so i tried to fixed it properly and rebased against latest master.

**Comment by kapouer** [comment]
The failure
```
server close should trigger on both ends upon ping timeout:
     Uncaught Error: expected 'transport error' to equal 'ping timeout'
```
happens on master as well, not related to this PR.

**Comment by alexhultman** [comment]
The ping failure is an old issue not related to uws. It happens with both ws and uws at random times due to too small timeout. Increase the timer for that case.

**Comment by kapouer** [comment]
Seems to pass, but let me remove this commit and make another PR just for that.

**Comment by alexhultman** [comment]
#461

**Comment by kapouer** [comment]
Yep :)

**Comment by darrachequesne** [comment]
@kapouer @alexhultman thanks!

**Comment by alexhultman** [comment]
Why is it an optional dependency? I heard this means it won't get downloaded by default and thus not really be used in most cases. Wouldn't it be better to make it a real dependency? Installation of uws cannot fail, it always succeeds silently. If someones server suddenly would blow up they could report an issue and use wsEngine: ws until it gets fixed.

**Comment by kapouer** [comment]
Hi,
this quote from npm's documentation explains it all: "The difference (for optionalDependencies) is that build failures do not cause installation to fail.".
So in *any* case `npm install engine.io` will try to install `uws`. Should it fail for any reason, it would fallback to the slow, horrible, scary native ws engine.

**Comment by alexhultman** [comment]
Oh, my bad then

**Comment by kapouer** [comment]
Pushing the reasoning further i would say `uws` could actually provide `ws` as a fallback in case it fails to install :)

**Comment by kapouer** [comment]
but you can consider this a new year's joke.

**Comment by alexhultman** [comment]
Even better, I could make it default to ws :+1:

**Comment by lpinca** [comment]
@kapouer 

> @darrachequesne the problem happened to be, IMO, in engine.io, so i tried to fixed it properly in 04bde4c and rebased against latest master.

I disagree, when you use

```js
const wss = new WebSocket.Server({ noServer: true });
```

you want an external server to be used and, in fact, `ws` works merely as an [`'upgrade'`](https://nodejs.org/dist/latest-v7.x/docs/api/http.html#http_event_upgrade_1) handler when `noServer` is `true`. Since there is no server, `wss.close()` only closes the open sockets.

On the other hand `uws` always uses its internal C++ server and if I'm not wrong when you call `wss.close()` you also close this C++ server even if `noServer` is `true`.

Now 2.0.0 is out and I think we should at least fix #473. Something like this works

```diff
diff --git a/lib/server.js b/lib/server.js
index 0da4c23..99dfb41 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -61,6 +61,8 @@ function Server (opts) {
       compression.threshold = 1024;
     }
   });
+
+  this.init();
 }
 
 /**
@@ -104,23 +106,24 @@ Server.prototype.clients;
  */
 
 Server.prototype.init = function () {
-  if (~this.transports.indexOf('websocket')) {
-    var wsModule;
-    try {
-      wsModule = require(this.wsEngine);
-    } catch (ex) {
-      this.wsEngine = 'ws';
-      // keep require('ws') as separate expression for packers (browserify, etc)
-      wsModule = require('ws');
-    }
-    var WebSocketServer = wsModule.Server;
-    this.ws = new WebSocketServer({
-      noServer: true,
-      clientTracking: false,
-      perMessageDeflate: this.perMessageDeflate,
-      maxPayload: this.maxHttpBufferSize
-    });
+  if (!~this.transports.indexOf('websocket')) return;
+
+  if (this.ws) this.ws.close();
+
+  var wsModule;
+  try {
+    wsModule = require(this.wsEngine);
+  } catch (ex) {
+    this.wsEngine = 'ws';
+    // keep require('ws') as separate expression for packers (browserify, etc)
+    wsModule = require('ws');
   }
+  this.ws = new wsModule.Server({
+    noServer: true,
+    clientTracking: false,
+    perMessageDeflate: this.perMessageDeflate,
+    maxPayload: this.maxHttpBufferSize
+  });
 };
 
 /**
```

**Comment by kapouer** [comment]
That's fine for me, as long as it works :)

**Comment by lpinca** [comment]
Another option would be to leave everything as is and specify that `Server.prototype.init()` should be called when [passing requests](https://github.com/socketio/engine.io#c-passing-in-requests).

```js
var engine = require('engine.io');
var server = new engine.Server();

server.init();

server.on('connection', function(socket){
  socket.send('hi');
});

// …
httpServer.on('upgrade', function(req, socket, head){
  server.handleUpgrade(req, socket, head);
});
httpServer.on('request', function(req, res){
  server.handleRequest(req, res);
});
```

**Pull request merged**